scylla/src/transport/session.rs

```diff
@@ -542,3 +604,39 @@ async fn resolve(addr: impl ToSocketAddrs + Display) -> Result<SocketAddr, Trans
 
     ret.ok_or(failed_err)
 }
+
+// "Keyspace names can have up to 48 alpha-numeric characters and contain underscores;
+// only letters and numbers are supported as the first character."
+// https://docs.datastax.com/en/cql-oss/3.3/cql/cql_reference/cqlCreateKeyspace.html
+// Despite that cassandra accepts undesrcore as first character so we do too
+// https://github.com/scylladb/scylla/blob/62551b3bd382c7c47371eb3fc38173bd0cfed44d/test/cql-pytest/test_keyspace.py#L58
+// https://github.com/scylladb/scylla/blob/fd1dd0eac7a303ceaa9f3ff643c4848506b0c85c/thrift/thrift_validation.cc#L76
+fn verify_keyspace_name_is_valid(keyspace_name: &str) -> Result<(), BadKeyspaceName> {
```

Comment:
An idea: let's define a [newtype](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) being a wrapper around a `str`/`String`/`Cow<'a, str>` etc. which carries information that the keyspace name inside it was verified. E.g.:

```rust
pub struct VerifiedKeyspaceName<'a>(&'a str);

impl<'a> VerifiedKeyspaceName<'a> {
    pub fn check(name: &'a str) -> Result<Self, BadKeyspaceName> {
        // do the actual verification here
        // ...
        Ok(VerifiedKeyspaceName(name))
    }

    pub fn as_str(&self) -> &str {
        self.0
    }
}
```

This way, you won't have to use comments in order to annotate that the keyspace name was verified. The compiler will guard you from passing an unverified name to a function which requires a verified name.

Reply:
Great idea, I added it